Nightly build failures, gcc/{11.3.0, 12.2.0, 13.2.0}, OpenMP+Serial backend: Kokkos_MemoryPool.hpp:169:45: error: invalid use of incomplete type 'using DefaultHostExecutionSpace = class Kokkos::OpenMP'

**Describe the bug**

A nightly Trilinos job with gcc/11.3.0 and the OpenMP backend enabled is failing to compile with output:
```
02:43:36 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_TaskQueue.hpp:33,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Serial/Kokkos_Serial_Task.hpp:27,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Serial/Kokkos_Serial.hpp:287,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_Default_Graph_Impl.hpp:26,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_Graph.hpp:219,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp: In member function 'void Kokkos::MemoryPool<Device>::get_usage_statistics(Kokkos::MemoryPool<Device>::usage_statistics&) const':
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp:169:45: error: invalid use of incomplete type 'using DefaultHostExecutionSpace = class Kokkos::OpenMP' {aka 'class Kokkos::OpenMP'}
02:43:36   169 |           Kokkos::DefaultHostExecutionSpace{}, sb_state_array, m_sb_state_array,
02:43:36       |                                             ^
02:43:36 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Build/kokkos/KokkosCore_Config_FwdBackend.hpp:22,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_Core_fwd.hpp:79,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_GraphImpl.hpp:22,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_Graph.hpp:31,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/fwd/Kokkos_Fwd_OPENMP.hpp:22:7: note: forward declaration of 'using DefaultHostExecutionSpace = class Kokkos::OpenMP' {aka 'class Kokkos::OpenMP'}
02:43:36    22 | class OpenMP;  ///< OpenMP execution space.
02:43:36       |       ^~~~~~
02:43:36 In file included from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_TaskQueue.hpp:33,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Serial/Kokkos_Serial_Task.hpp:27,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Serial/Kokkos_Serial.hpp:287,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/impl/Kokkos_Default_Graph_Impl.hpp:26,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_Graph.hpp:219,
02:43:36                  from /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp: In member function 'void Kokkos::MemoryPool<Device>::print_state(std::ostream&) const':
02:43:36 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/Trilinos/kokkos/core/src/Kokkos_MemoryPool.hpp:222:45: error: invalid use of incomplete type 'using DefaultHostExecutionSpace = class Kokkos::OpenMP' {aka 'class Kokkos::OpenMP'}
...
```

This follows after merge of these changes:

**Changes**

    Print value of all relevant SYCL_EXT_ONEAPI* macros ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/440/changes#detail))
    Always pass execution space to `Impl::DeepCopy` (#7519) ([detail](https://jenkins-son.sandia.gov/job/KokkosEco_Trilinos_Blake_Gcc_1130_OpenMPI_415_OPENMP-4t/440/changes#detail))

Presumably correlated with #7519

I'll test a similar standalone Kokkos build to see about reproducing outside of Trilinos

**Please include the following for a minimal reproducer**

**Reproducer (with Trilinos)**: (blake all queue)
```bash
# Repo prep
git clone -b develop https://github.com/trilinos/Trilinos.git
TRILINOS_DIR=$PWD/Trilinos
git clone -b develop https://github.com/kokkos/kokkos.git
KOKKOS_DIR=$PWD/kokkos
git clone -b develop https://github.com/kokkos/kokkos-kernels.git
KOKKOSKERNELS_DIR=$PWD/kokkos-kernels

cd $TRILINOS_DIR
ln -s ${KOKKOS_DIR}/kokkos kokkos
ln -s ${KOKKOSKERNELS_DIR}/kokkos-kernels kokkos-kernels
cd -

mkdir -p Build
cd Build


salloc -N 1 -p all

# Environment
module purge
module load cmake gcc/11.3.0 openmpi/4.1.5 openblas/0.3.23

# Configure Trilinos - likely only Kokkos needs to be enabled
cmake \
  -D CMAKE_CXX_COMPILER="`which mpicxx`" \
  -D CMAKE_C_COMPILER="`which mpicc`" \
  -D CMAKE_Fortran_COMPILER="`which mpifort`" \
  -D CMAKE_CXX_FLAGS="-g" \
  -D CMAKE_C_FLAGS="-g" \
  -DTPL_ENABLE_MPI=ON \
  -DTPL_ENABLE_BLAS:BOOL=ON \
   -DBLAS_LIBRARY_DIRS:FILEPATH=${OPENBLAS_ROOT}/lib \
   -DBLAS_LIBRARY_NAMES:STRING="openblas" \
  -DTPL_ENABLE_LAPACK:BOOL=ON \
   -DLAPACK_INCLUDE_DIRS:FILEPATH=${OPENBLAS_ROOT}/include \
   -DLAPACK_LIBRARY_DIRS:FILEPATH=${OPENBLAS_ROOT}/lib \
   -DLAPACK_LIBRARY_NAMES:STRING="openblas" \
  -DTrilinos_ENABLE_ALL_PACKAGES=OFF \
  -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES=OFF \
  -DTrilinos_ENABLE_TESTS=ON \
  -DTrilinos_MUST_FIND_ALL_TPL_LIBS=TRUE \
  -DTrilinos_ENABLE_COMPLEX=ON \
  -DTrilinos_ENABLE_OpenMP=ON \
  -DTrilinos_ENABLE_Kokkos=ON \
  -D Kokkos_ENABLE_OPENMP=ON \
   -D Kokkos_ENABLE_TESTS=ON \
  -D Kokkos_ARCH_SKX=ON \
   -D Kokkos_ENABLE_IMPL_MDSPAN=ON \
  -DTrilinos_ENABLE_KokkosKernels=ON \
   -D KokkosKernels_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Tpetra=ON \
   -D Tpetra_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Sacado=ON \
   -D Sacado_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Stokhos=ON \
   -D Stokhos_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Intrepid2=ON \
   -D Intrepid2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Amesos2=ON \
   -D Amesos2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Ifpack2=ON \
   -D Ifpack2_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Zoltan2=ON \
   -D Zoltan2_ENABLE_TESTS=ON \
\
  -DTPL_ENABLE_Matio=OFF \
\
-DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos \
-DKokkosKernels_SOURCE_DIR_OVERRIDE:STRING=kokkos-kernels \
$TRILINOS_DIR
```

Reproducer (kokkos only):
```bash
module load cmake gcc/11.3.0 openmpi/4.1.5 openblas/0.3.23

cmake \
  -D CMAKE_CXX_COMPILER="g++" \
  -D CMAKE_CXX_FLAGS="-g" \
  -D CMAKE_BUILD_TYPE="RELEASE" \
  -D Kokkos_ENABLE_SERIAL=ON \
  -D Kokkos_ENABLE_OPENMP=ON \
  -D Kokkos_ENABLE_TESTS=ON \
  -D Kokkos_ARCH_SKX=ON \
  -D Kokkos_ENABLE_IMPL_MDSPAN=ON \
$KOKKOS_DIR
```

* The explicit enable of both the OPENMP and SERIAL backends was needed to trigger the error; only enabling the OPENMP backend allowed compilation to succeed

Edit: added standalone kokkos reproducer infor

--- a/core/src/Kokkos_MemoryPool.hpp
+++ b/core/src/Kokkos_MemoryPool.hpp
@@ -157,8 +157,13 @@ class MemoryPool {
     size_t reserved_bytes;   ///<  Unallocated bytes in assigned superblocks
   };
 
+  // This function is templated to avoid needing a full definition of
+  // DefaultHostExecutionSpace at class instantiation
+  template <typename ExecutionSpace = Kokkos::DefaultHostExecutionSpace>
   void get_usage_statistics(usage_statistics &stats) const {
     Kokkos::HostSpace host;
+    static_assert(
+        std::is_same_v<ExecutionSpace, Kokkos::DefaultHostExecutionSpace>);
 
     const size_t alloc_size = m_hint_offset * sizeof(uint32_t);
 
@@ -167,8 +172,7 @@ class MemoryPool {
 
     if (!accessible) {
       Kokkos::Impl::DeepCopy<Kokkos::HostSpace, base_memory_space>(
-          Kokkos::DefaultHostExecutionSpace{}, sb_state_array, m_sb_state_array,
-          alloc_size);
+          ExecutionSpace{}, sb_state_array, m_sb_state_array, alloc_size);
       Kokkos::fence(
           "MemoryPool::get_usage_statistics(): fence after copying state "
           "array to HostSpace");
@@ -210,8 +214,13 @@ class MemoryPool {
     }
   }
 
+  // This function is templated to avoid needing a full definition of
+  // DefaultHostExecutionSpace at class instantiation
+  template <typename ExecutionSpace = Kokkos::DefaultHostExecutionSpace>
   void print_state(std::ostream &s) const {
     Kokkos::HostSpace host;
+    static_assert(
+        std::is_same_v<ExecutionSpace, Kokkos::DefaultHostExecutionSpace>);
 
     const size_t alloc_size = m_hint_offset * sizeof(uint32_t);
 
@@ -220,8 +229,7 @@ class MemoryPool {
 
     if (!accessible) {
       Kokkos::Impl::DeepCopy<Kokkos::HostSpace, base_memory_space>(
-          Kokkos::DefaultHostExecutionSpace{}, sb_state_array, m_sb_state_array,
-          alloc_size);
+          ExecutionSpace{}, sb_state_array, m_sb_state_array, alloc_size);
       Kokkos::fence(
           "MemoryPool::print_state(): fence after copying state array to "
           "HostSpace");

